presto-main/src/main/java/com/facebook/presto/operator/OrderByOperator.java

```diff
@@ -185,23 +213,72 @@ public Page getOutput()
             return null;
         }
 
-        if (currentPosition >= pageIndex.getPositionCount()) {
+        if (!mergedPages.process() || mergedPages.isFinished()) {
             state = State.FINISHED;
             return null;
         }
 
-        // iterate through the positions sequentially until we have one full page
-        pageBuilder.reset();
-        currentPosition = pageIndex.buildPage(currentPosition, outputChannels, pageBuilder);
+        Page resultPage = mergedPages.getResult();
+        Block[] blocksArray = new Block[outputChannels.length];
 
-        // output the page if we have any data
-        if (pageBuilder.isEmpty()) {
-            state = State.FINISHED;
-            return null;
+        // Do not use revocable memory during unspill
+        updateMemoryUsage(false);
+
+        for (int i = 0; i < outputChannels.length; i++) {
+            blocksArray[i] = resultPage.getBlock(outputChannels[i]);
+        }
+
+        return new Page(resultPage.getPositionCount(), blocksArray);
+    }
+
+    @Override
+    public ListenableFuture<?> startMemoryRevoke()
+    {
+        return spillToDisk();
+    }
+
+    public ListenableFuture<?> spillToDisk()
+    {
+        checkState(hasPreviousSpillCompletedSuccessfully(), "Previous spill hasn't yet finished");
+
+        if (!spiller.isPresent()) {
+            this.spiller = Optional.of(spillerFactory.get().create(
+                    sourceTypes,
+                    operatorContext.getSpillContext(),
+                    operatorContext.newAggregateSystemMemoryContext()));
+        }
+
+        if (pageIndex.getPositionCount() == 0) {
+            return spillInProgress;
+        }
+
+        pageIndex.sort(sortChannels, sortOrder);
+        spillInProgress = spiller.get().spill(pageIndex.getSortedPages());
+
+        return spillInProgress;
+    }
+
+    private boolean hasPreviousSpillCompletedSuccessfully()
+    {
+        if (spillInProgress.isDone()) {
+            // check for exception from previous spill for early failure
+            getFutureValue(spillInProgress);
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    @Override
+    public void finishMemoryRevoke()
+    {
+        if (pageIndex.getPositionCount() == 0) {
+            return;
         }
 
-        Page page = pageBuilder.build();
-        return page;
+        pageIndex.clear();
+        localRevocableMemoryContext.setBytes(pageIndex.getEstimatedSize().toBytes());
```

Comment:
`pageIndex` memory overhead is non-revokable, see: https://github.com/prestodb/presto/pull/11364/files#diff-f0a46746520bef39a6a15ddb4e65aa56R607

Reply:
Good catch, thanks